File: app/src/main/java/com/gzuschedule/app/data/update/UpdateChecker.kt
package com.gzuschedule.app.data.update

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

/**
 * 检查更新（ADR-107）。
 *
 * ⚠️ 用户需求：
 *    「在设置里加入检查更新的功能 自动访问我的仓库看看软件有没有新版本
 *      可以设置开启或者关闭」
 *
 * 实现：
 *   · 请求 GitHub Releases API（最新的 release）
 *   · 与本机 versionName 比对
 *   · 只读，不需要 token（仓库是 public）
 *
 * ⚠️ 隐私：只发一个 GET，不带任何用户信息。
 */
object UpdateChecker {

    /** 仓库坐标。 */
    private const val OWNER = "xiaoxiao602"
    private const val REPO = "ScheduleApp"

    /** GitHub Releases API（取最新一个非 draft / 非 prerelease）。 */
    private const val API = "https://api.github.com/repos/$OWNER/$REPO/releases/latest"

    /** 超时短一点 —— 检查更新不该让用户等。 */
    private val client: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(6, TimeUnit.SECONDS)
            .readTimeout(8, TimeUnit.SECONDS)
            .build()
    }

    private val gson = Gson()

    /** 检查结果。 */
    data class Result(
        /** 有新版可用。 */
        val hasUpdate: Boolean,
        /** 最新版本号（去掉前缀 v），如 "1.0.2"。 */
        val latestVersion: String?,
        /** Release 页面地址（让用户去下载）。 */
        val releaseUrl: String?,
        /** APK 直链（可能为空）。 */
        val apkUrl: String?,
        /** 更新说明（Release body）。 */
        val notes: String?,
        /** 出错时的信息（无错为 null）。 */
        val error: String? = null,
    )

    /**
     * 检查是否有新版本。
     *
     * @param currentVersion 当前 versionName，如 "1.0.1"
     */
    suspend fun check(currentVersion: String): Result = withContext(Dispatchers.IO) {
        runCatching {
            val req = Request.Builder()
                .url(API)
                .header("Accept", "application/vnd.github+json")
                .header("User-Agent", "GzuScheduleApp")
                .build()

            client.newCall(req).execute().use { resp ->
                if (!resp.isSuccessful) {
                    return@runCatching Result(
                        false, null, null, null, null,
                        error = "服务器返回 ${resp.code}",
                    )
                }
                val body = resp.body?.string()
                    ?: return@runCatching Result(
                        false, null, null, null, null, error = "响应为空",
                    )

                val rel = gson.fromJson(body, ReleaseJson::class.java)
                    ?: return@runCatching Result(
                        false, null, null, null, null, error = "解析失败",
                    )

                val latest = rel.tagName?.trim()?.removePrefix("v")?.removePrefix("V")
                    ?: return@runCatching Result(
                        false, null, null, null, null, error = "无版本号",
                    )

                val apk = rel.assets
                    ?.firstOrNull { a -> a.name?.endsWith(".apk", ignoreCase = true) == true }
                    ?.browserDownloadUrl

                Result(
                    hasUpdate = isNewer(latest, currentVersion),
                    latestVersion = latest,
                    releaseUrl = rel.htmlUrl,
                    apkUrl = apk,
                    notes = rel.body?.take(600),
                )
            }
        }.getOrElse { e ->
            Result(false, null, null, null, null, error = e.message ?: "网络错误")
        }
    }

    /**
     * 判断 [latest] 是否比 [current] 新。
     *
     * ⚠️ 用**逐段数字比较**，不能直接字符串比（"1.0.10" < "1.0.9" 是错的）。
     *    非数字段按 0 处理（如 "1.0.1-beta" → [1,0,1]）。
     */
    fun isNewer(latest: String, current: String): Boolean {
        val a = parseVersion(latest)
        val b = parseVersion(current)
        val n = maxOf(a.size, b.size)
        for (i in 0 until n) {
            val x = a.getOrElse(i) { 0 }
            val y = b.getOrElse(i) { 0 }
            if (x != y) return x > y
        }
        return false   // 完全相同
    }

    /** "1.0.1" → [1, 0, 1]；取不出数字的段按 0。 */
    private fun parseVersion(v: String): List<Int> =
        v.split(".", "-", "_", "+")
            .map { seg ->
                seg.takeWhile { it.isDigit() }.toIntOrNull() ?: 0
            }

    // ---------- GitHub API 响应模型 ----------

    private data class ReleaseJson(
        @SerializedName("tag_name") val tagName: String?,
        @SerializedName("html_url") val htmlUrl: String?,
        @SerializedName("body") val body: String?,
        val assets: List<AssetJson>?,
    )

    private data class AssetJson(
        val name: String?,
        @SerializedName("browser_download_url") val browserDownloadUrl: String?,
    )
}

File: lib/domain/update_checker.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'core_logic.dart' show VersionCompare;

/// ============================================================
/// 检查更新（对应原版 data/update/UpdateChecker.kt）
/// ============================================================
///
/// 数据源：GitHub Releases API
///   GET https://api.github.com/repos/{owner}/{repo}/releases/latest
///   Header: Accept: application/vnd.github+json
///
/// 比对本机 versionName 与 tag_name，得出是否有新版。
///
/// ⚠️ 本仓库是 public —— 无需 token。GitHub 对匿名请求限流 60 次/小时，
///    对"用户手动点一次"完全够用。
/// ⚠️ 不要在这里做"自动下载安装" —— 应用商店/系统权限都不允许静默安装；
///    原版也只是打开浏览器让用户自己下。

/// 一次检查的结果。
class UpdateInfo {
  const UpdateInfo({
    required this.hasUpdate,
    required this.currentVersion,
    required this.latestVersion,
    this.releaseUrl,
    this.downloadUrl,
    this.notes,
    this.publishedAt,
  });

  final bool hasUpdate;
  final String currentVersion;
  final String latestVersion;

  /// Release 页面（用户点「去下载」时打开这个）。
  final String? releaseUrl;

  /// 直接的 APK 下载地址（Release 里的 asset）。
  final String? downloadUrl;

  final String? notes;
  final String? publishedAt;
}

abstract final class UpdateChecker {
  /// ⚠️ 改成你的仓库坐标。
  static const owner = 'xiaoxiao602';
  static const repo = 'ScheduleApp';

  static Uri get _api =>
      Uri.parse('https://api.github.com/repos/$owner/$repo/releases/latest');

  /// 检查更新。[currentVersion] 形如 "1.2.0"。
  static Future<UpdateInfo> check(String currentVersion) async {
    final resp = await http.get(
      _api,
      headers: const {
        'Accept': 'application/vnd.github+json',
        'User-Agent': 'GzuSchedule/Flutter',
      },
    ).timeout(const Duration(seconds: 15));

    if (resp.statusCode != 200) {
      throw Exception('GitHub API ${resp.statusCode}');
    }

    final json = jsonDecode(utf8.decode(resp.bodyBytes)) as Map<String, dynamic>;
    final tag = (json['tag_name'] as String? ?? '').replaceAll('v', '').trim();
    final htmlUrl = json['html_url'] as String?;
    final body = json['body'] as String?;
    final published = json['published_at'] as String?;

    // 从 assets 里找 .apk
    String? apk;
    final assets = json['assets'];
    if (assets is List) {
      for (final a in assets) {
        if (a is Map && (a['name'] as String? ?? '').toLowerCase().endsWith('.apk')) {
          apk = a['browser_download_url'] as String?;
          break;
        }
      }
    }

    return UpdateInfo(
      hasUpdate: tag.isNotEmpty && VersionCompare.isNewer(tag, currentVersion),
      currentVersion: currentVersion,
      latestVersion: tag,
      releaseUrl: htmlUrl,
      downloadUrl: apk,
      notes: body,
      publishedAt: published,
    );
  }
}
